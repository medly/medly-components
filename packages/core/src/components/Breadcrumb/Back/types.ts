import { HTMLProps } from '@medly-components/utils';

export type BreadcrumbBackProps = HTMLProps<HTMLLIElement> & {
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
};
