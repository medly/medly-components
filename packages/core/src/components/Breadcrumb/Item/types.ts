import { HTMLProps } from '@medly-components/utils';

export interface BreadcrumbItemProps extends HTMLProps<HTMLLIElement> {
    /** To be used to disable Item  */
    disabled?: boolean;
}
