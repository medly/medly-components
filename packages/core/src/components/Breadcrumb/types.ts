import { HTMLProps } from '@medly-components/utils';
import type { FC, ReactNode } from 'react';

export type BreadcrumbProps = Omit<HTMLProps<HTMLOListElement>, 'type'> & {
    /** You can pass any separator which you can use between the links */
    separator?: string | ReactNode;
};

export type BreadcrumbStaticProps = {
    Item: FC<HTMLProps<HTMLLIElement>>;
    Back: FC<HTMLProps<HTMLLIElement>>;
};
