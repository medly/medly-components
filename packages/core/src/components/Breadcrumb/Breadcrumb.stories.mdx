import { Breadcrumb } from './Breadcrumb.tsx';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import * as stories from './Breadcrumb.stories';
import { action } from '@storybook/addon-actions';
import Text from '../Text';

<Meta title="Core" component={Breadcrumb} />

# Breadcrumb

Breadcrumbs indicate current page's location to user.

<Preview withToolbar>
    <Story name="Breadcrumb">
        <Breadcrumb>
            <Breadcrumb.Item onClick={action('Home')}>Home</Breadcrumb.Item>
            <Breadcrumb.Item onClick={action('Article')}>Article</Breadcrumb.Item>
            <Breadcrumb.Item disabled onClick={action('Section')}>
                Section
            </Breadcrumb.Item>
        </Breadcrumb>
    </Story>
</Preview>

You have to pass `onClick` props to perform any action on click on any item. And you can pass `disabled` prop to disable any item.

Also you can use back button by calling `<Breadcrumb.Back />`

<Preview withToolbar>
    <Breadcrumb.Back onClick={action('Back')} />
</Preview>

### Props

<Props of={Breadcrumb} />

### BreadcrumbTheme

<Props of={stories.ThemeInterface} />
