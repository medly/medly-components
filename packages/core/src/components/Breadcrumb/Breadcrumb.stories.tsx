import { BreadcrumbTheme, defaultTheme } from '@medly-components/theme';
import type { FC } from 'react';

export const ThemeInterface: FC<BreadcrumbTheme> = () => null;
ThemeInterface.defaultProps = {
    ...defaultTheme.breadcrumb
};
