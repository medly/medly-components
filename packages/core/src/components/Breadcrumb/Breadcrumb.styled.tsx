import { SvgIcon } from '@medly-components/icons';
import styled from 'styled-components';

export const BreadcrumbStyled = styled('ol')`
    height: 2.4rem;
    display: flex;
    align-items: center;
    padding: 0;
    margin: 0;
    list-style: none;
    flex-direction: row;

    & > *:nth-child(even) {
        margin: 0 0.4rem;
    }

    & > ${SvgIcon} {
        * {
            fill: ${({ theme }) => theme.breadcrumb.textColor.default};
        }
    }
`;
