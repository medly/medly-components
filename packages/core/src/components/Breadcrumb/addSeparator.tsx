import { isValidStringOrNumber } from '@medly-components/utils';
import Text from '../Text';
import { BreadcrumbProps } from './types';

const addSeparator = (links: any[], separator: BreadcrumbProps['separator']) =>
    links.reduce(
        (acc: any, cur, index) => [
            ...acc,
            cur,
            ...(index !== links.length - 1
                ? [
                      isValidStringOrNumber(separator) ? (
                          <Text key={index} aria-hidden="true">
                              {separator}
                          </Text>
                      ) : (
                          separator
                      )
                  ]
                : [])
        ],
        []
    );

export default addSeparator;
