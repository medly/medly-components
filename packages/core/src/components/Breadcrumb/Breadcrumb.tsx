import { ChevronRightIcon } from '@medly-components/icons';
import { WithStyle } from '@medly-components/utils';
import { Children } from 'react';
import addSeparator from './addSeparator';
import BreadcrumbBack from './Back';
import { BreadcrumbStyled } from './Breadcrumb.styled';
import BreadcrumbItem from './Item';
import { BreadcrumbProps, BreadcrumbStaticProps } from './types';
import type { FC } from 'react';

const Component: FC<BreadcrumbProps> = props => {
    const { separator, children, ...restProps } = props,
        links = Children.toArray(children),
        linksWithSeparator = addSeparator(links, separator);

    return <BreadcrumbStyled {...restProps}>{linksWithSeparator}</BreadcrumbStyled>;
};

Component.displayName = 'Breadcrumb';
Component.defaultProps = { separator: <ChevronRightIcon size="S" aria-hidden="true" /> };
export const Breadcrumb: FC<BreadcrumbProps> & WithStyle & BreadcrumbStaticProps = Object.assign(Component, {
    Item: BreadcrumbItem,
    Back: BreadcrumbBack,
    Style: BreadcrumbStyled
});
