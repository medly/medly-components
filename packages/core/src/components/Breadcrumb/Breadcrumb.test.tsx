import { ChevronRightIcon } from '@medly-components/icons';
import { render } from '@test-utils';
import Text from '../Text';
import { Breadcrumb } from './Breadcrumb';
import { BreadcrumbProps } from './types';

describe('Breadcrumb component', () => {
    test.each(['/', <ChevronRightIcon size="S" key="dummy" />])(
        'should render properly with %p separator',
        (separator: BreadcrumbProps['separator']) => {
            const { container } = render(
                <Breadcrumb separator={separator}>
                    <Breadcrumb.Back />
                    <Breadcrumb.Item>
                        <Text>Covered Entities</Text>
                    </Breadcrumb.Item>
                    <Breadcrumb.Item onClick={jest.fn()}>Add New Entities</Breadcrumb.Item>
                </Breadcrumb>
            );
            expect(container).toMatchSnapshot();
        }
    );
});
