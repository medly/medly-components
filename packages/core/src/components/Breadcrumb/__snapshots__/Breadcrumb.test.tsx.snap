// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb component should render properly with "/" separator 1`] = `
.c3 {
  overflow: visible;
  font-size: 2.4rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: inherit;
}

.c3 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c6 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c6 .c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #3872D2;
  font-size: 1.4rem;
  font-weight: 600;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

.c6:hover .c4 {
  color: #3061B3;
}

.c7 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c7 .c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #3872D2;
  font-size: 1.4rem;
  font-weight: 600;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

.c7:hover .c4 {
  color: #3061B3;
}

.c1 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 .c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #3872D2;
  font-size: 1.4rem;
  font-weight: 600;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

.c1:hover .c4 {
  color: #3061B3;
}

.c1 > .c2 {
  margin-right: 0.4rem;
}

.c1 > .c2 * {
  fill: #3872D2;
}

.c0 {
  height: 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0;
  margin: 0;
  list-style: none;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 > *:nth-child(even) {
  margin: 0 0.4rem;
}

.c0 > .c2 * {
  fill: #3872D2;
}

<div>
  <ol
    class="c0"
  >
    <li
      class="c1"
    >
      <svg
        class="c2 c3"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15 6.705a.996.996 0 0 0-1.41 0L9 11.295a.996.996 0 0 0 0 1.41l4.59 4.59a.996.996 0 1 0 1.41-1.41l-3.88-3.89L15 8.115c.39-.39.38-1.03 0-1.41z"
          fill="#000"
          fill-opacity=".54"
        />
      </svg>
      <span
        class="c4 c5"
      >
        Back
      </span>
    </li>
    <span
      aria-hidden="true"
      class="c4 c5"
    >
      /
    </span>
    <li
      class="c6"
    >
      <span
        class="c4 c5"
      >
        Covered Entities
      </span>
    </li>
    <span
      aria-hidden="true"
      class="c4 c5"
    >
      /
    </span>
    <li
      class="c7"
    >
      <span
        class="c4 c5"
      >
        Add New Entities
      </span>
    </li>
  </ol>
</div>
`;

exports[`Breadcrumb component should render properly with {"$typeof": Symbol(react.element), "_owner": null, "_store": [Object], "key": "dummy", "props": [Object], "ref": null, "type": [Object]} separator 1`] = `
.c3 {
  overflow: visible;
  font-size: 2.4rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: inherit;
}

.c3 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c6 {
  overflow: visible;
  font-size: 2rem;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  cursor: inherit;
}

.c6 * {
  fill-opacity: 1;
  -webkit-transition: all 100ms linear;
  transition: all 100ms linear;
  fill: #607890;
}

.c5 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

.c7 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7 .c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #3872D2;
  font-size: 1.4rem;
  font-weight: 600;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

.c7:hover .c4 {
  color: #3061B3;
}

.c8 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c8 .c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #3872D2;
  font-size: 1.4rem;
  font-weight: 600;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

.c8:hover .c4 {
  color: #3061B3;
}

.c1 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1 .c4 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: #3872D2;
  font-size: 1.4rem;
  font-weight: 600;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  font-family: Open Sans,Helvetica Neue,Helvetica,Arial,sans-serif;
}

.c1:hover .c4 {
  color: #3061B3;
}

.c1 > .c2 {
  margin-right: 0.4rem;
}

.c1 > .c2 * {
  fill: #3872D2;
}

.c0 {
  height: 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0;
  margin: 0;
  list-style: none;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 > *:nth-child(even) {
  margin: 0 0.4rem;
}

.c0 > .c2 * {
  fill: #3872D2;
}

<div>
  <ol
    class="c0"
  >
    <li
      class="c1"
    >
      <svg
        class="c2 c3"
        fill="none"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15 6.705a.996.996 0 0 0-1.41 0L9 11.295a.996.996 0 0 0 0 1.41l4.59 4.59a.996.996 0 1 0 1.41-1.41l-3.88-3.89L15 8.115c.39-.39.38-1.03 0-1.41z"
          fill="#000"
          fill-opacity=".54"
        />
      </svg>
      <span
        class="c4 c5"
      >
        Back
      </span>
    </li>
    <svg
      class="c2 c6"
      fill="none"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 6.71a.996.996 0 0 0 0 1.41L12.88 12 9 15.88a.996.996 0 1 0 1.41 1.41L15 12.7a.996.996 0 0 0 0-1.41L10.41 6.7c-.38-.38-1.02-.38-1.41.01z"
        fill="#000"
        fill-opacity=".54"
      />
    </svg>
    <li
      class="c7"
    >
      <span
        class="c4 c5"
      >
        Covered Entities
      </span>
    </li>
    <svg
      class="c2 c6"
      fill="none"
      height="1em"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 6.71a.996.996 0 0 0 0 1.41L12.88 12 9 15.88a.996.996 0 1 0 1.41 1.41L15 12.7a.996.996 0 0 0 0-1.41L10.41 6.7c-.38-.38-1.02-.38-1.41.01z"
        fill="#000"
        fill-opacity=".54"
      />
    </svg>
    <li
      class="c8"
    >
      <span
        class="c4 c5"
      >
        Add New Entities
      </span>
    </li>
  </ol>
</div>
`;
